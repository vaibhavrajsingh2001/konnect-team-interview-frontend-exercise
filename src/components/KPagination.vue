<template>
  <div class="paginator">
    <button
      :disabled="disablePreviousButton"
      @click="$emit('previous')"
    >
      &lt;—
    </button>
    <span>{{ firstItemIndex }} to {{ lastItemIndex }} of {{ totalCount }} services</span>
    <button
      :disabled="disableNextButton"
      @click="$emit('next')"
    >
      —&gt;
    </button>
  </div>
</template>

<script lang="ts">
import { defineComponent } from 'vue'

export default defineComponent({
  name: 'KPagination',
  props: {
    totalPages: {
      type: Number,
      required: true,
    },
    currentPage: {
      type: Number,
      required: true,
    },
    pageSize: {
      type: Number,
      required: true,
    },
    totalCount: {
      type: Number,
      required: true,
    },
  },
  emits: ['next', 'previous'],
  computed: {
    // Return the index of the first item on the current page
    firstItemIndex() {
      /**
       * In case pageSize === 10
       * - If the current page is 1, the first item is 1
       * - If the current page is 2, the first item is 11
       *  and so on
       */
      return ((this.currentPage - 1) * this.pageSize) + 1

    },
    // Return the index of the last item on the current page
    lastItemIndex() {
      // If there is only one page, the total count is the last item index
      // e.g. if there are only 6 items, then only 1 page will be shown, so the last item index will be 6
      if (this.totalPages === 1) return this.totalCount

      /**
       * In case pageSize === 10
       * - If the current page is 1, the last item is 10
       * - If the current page is 2, the last item is 20
       *  and so on
       */
      return this.currentPage * this.pageSize
    },
    disablePreviousButton() {
      return this.currentPage <= 1
    },
    disableNextButton() {
      return this.currentPage >= this.totalPages
    },
  },
})
</script>

<style lang="scss">
.paginator {
  text-align: center;
}
</style>
