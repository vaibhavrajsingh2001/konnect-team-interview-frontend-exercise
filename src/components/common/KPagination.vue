<template>
  <div class="k-pagination">
    <button
      :disabled="disablePreviousButton"
      @click="$emit('previous')"
    >
      ←
    </button>

    <span class="current-page">
      {{ firstItemIndex }} to {{ lastItemIndex }}
      <span class="total-count">of {{ totalCount }} services</span>
    </span>

    <button
      :disabled="disableNextButton"
      @click="$emit('next')"
    >
      →
    </button>
  </div>
</template>

<script lang="ts">
import { defineComponent } from 'vue'

export default defineComponent({
  name: 'KPagination',
  props: {
    totalPages: {
      type: Number,
      required: true,
    },
    currentPage: {
      type: Number,
      required: true,
    },
    pageSize: {
      type: Number,
      required: true,
    },
    totalCount: {
      type: Number,
      required: true,
    },
  },
  emits: ['next', 'previous'],
  computed: {
    // Return the index of the first item on the current page
    firstItemIndex() {
      /**
       * In case pageSize === 10
       * - If the current page is 1, the first item is 1
       * - If the current page is 2, the first item is 11
       *  and so on
       */
      return ((this.currentPage - 1) * this.pageSize) + 1

    },
    // Return the index of the last item on the current page
    lastItemIndex() {
      // If there is only one page, the total count is the last item index
      // e.g. if there are only 6 items, then only 1 page will be shown, so the last item index will be 6
      if (this.totalPages === 1) return this.totalCount

      // If the current page is the last page, the total count is the last item index
      if (this.currentPage === this.totalPages) return this.totalCount

      /**
       * In case pageSize === 10
       * - If the current page is 1, the last item is 10
       * - If the current page is 2, the last item is 20
       *  and so on
       */
      return this.currentPage * this.pageSize
    },
    disablePreviousButton() {
      return this.currentPage <= 1
    },
    disableNextButton() {
      return this.currentPage >= this.totalPages
    },
  },
})
</script>

<style lang="scss">
.k-pagination {
  align-items: center;
  display: flex;
  font-size: 1.3rem;
  gap: 2.4rem;
  justify-content: center;
  line-height: 2rem;

  button {
    background-color: transparent;
    border: 1px solid #A6C6FF;
    border-radius: 50%;
    color: #1456CB;
    cursor: pointer;
    font-size: 1.8rem;
    padding: 1.2rem;

    &:disabled {
      border: 1px solid #0000001A;
      color: #0000001A;
      cursor: not-allowed;
    }
  }

  .current-page {
    color: #3C4557;
    font-weight: 600;
    text-align: center;
    // Fixing the width so that the text doesn't jump around
    // e.g. going from "1 to 10 of 100 services" to "11 to 20 of 100 services" would cause the text to jump
    width: 14.5rem;

    .total-count {
      color: #707888;
      font-weight: 500;
    }
  }

}
</style>
