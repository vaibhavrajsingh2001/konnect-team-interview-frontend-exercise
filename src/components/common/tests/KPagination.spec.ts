import { mount, renderToString } from '@vue/test-utils'
import { describe, it, expect } from 'vitest'
import KPagination from '@/components/common/KPagination.vue'

describe('KPagination', () => {
  it('renders correctly', async () => {
    const wrapper = await renderToString(KPagination, {
      props: {
        totalPages: 5,
        currentPage: 2,
        pageSize: 10,
        totalCount: 50,
      },
    })
    expect(wrapper).toMatchSnapshot()
  })

  it('renders the correct page range and total count', () => {
    const wrapper = mount(KPagination, {
      props: {
        totalPages: 5,
        currentPage: 2,
        pageSize: 10,
        totalCount: 50,
      },
    })

    expect(wrapper.findTestId('current-page-label').text()).toBe('11 to 20 of 50 services')
  })

  it('disables the previous button on the first page', () => {
    const wrapper = mount(KPagination, {
      props: {
        totalPages: 5,
        currentPage: 1,
        pageSize: 10,
        totalCount: 50,
      },
    })

    expect(wrapper.findTestId<'button'>('previous-button').attributes().disabled).toBeDefined()
  })

  it('disables the next button on the last page', () => {
    const wrapper = mount(KPagination, {
      props: {
        totalPages: 5,
        currentPage: 5,
        pageSize: 10,
        totalCount: 50,
      },
    })

    expect(wrapper.findTestId<'button'>('next-button').attributes().disabled).toBeDefined()
  })

  it('enables both previous and next buttons on a middle page', () => {
    const wrapper = mount(KPagination, {
      props: {
        totalPages: 5,
        currentPage: 3,
        pageSize: 10,
        totalCount: 50,
      },
    })

    expect(wrapper.findTestId<'button'>('previous-button').attributes().disabled).toBeUndefined()
    expect(wrapper.findTestId<'button'>('next-button').attributes().disabled).toBeUndefined()
  })

  it('displays the correct last item index when there is only one page', () => {
    const wrapper = mount(KPagination, {
      props: {
        totalPages: 1,
        currentPage: 1,
        pageSize: 10,
        totalCount: 6,
      },
    })

    expect(wrapper.findTestId('current-page-label').text()).toBe('1 to 6 of 6 services')
  })

  it('displays the correct last item index on the last page', () => {
    const wrapper = mount(KPagination, {
      props: {
        totalPages: 5,
        currentPage: 5,
        pageSize: 10,
        totalCount: 50,
      },
    })

    expect(wrapper.findTestId('current-page-label').text()).toBe('41 to 50 of 50 services')
  })

  it('emits the correct events when the buttons are clicked', async () => {
    const wrapper = mount(KPagination, {
      props: {
        totalPages: 5,
        currentPage: 3,
        pageSize: 10,
        totalCount: 50,
      },
    })

    await wrapper.findTestId<'button'>('next-button').trigger('click')
    expect(wrapper.emitted()).toHaveProperty('next')

    await wrapper.findTestId<'button'>('previous-button').trigger('click')
    expect(wrapper.emitted()).toHaveProperty('previous')
  })
})
