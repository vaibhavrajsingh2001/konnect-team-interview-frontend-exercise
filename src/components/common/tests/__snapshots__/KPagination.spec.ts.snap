// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`KPagination > renders correctly 1`] = `"<div class="k-pagination"><button data-testid="previous-button"> ← </button><span class="current-page" data-testid="current-page-label">11 to 20 <span class="total-count">of 50 services</span></span><button data-testid="next-button"> → </button></div>"`;
